Add libretro VFS support

garbear opened the pull request:
## Description

This PR adds support for version 1 of the libretro VFS API (https://github.com/libretro/RetroArch/pull/5664). It's a straightforward translation between the libretro side and the Kodi side.

## How has this been tested?

Time is short, so I haven't started testing yet. It compiles against Matrix and visibly appears correct. I'm PRing now to get some visibility and keep track of this task for later.

garbear committed:
Add libretro VFS support (v1)

AlwinEsch commented on src/libretro/FrontendBridge.cpp:
```diff
@@ -345,3 +347,178 @@ void CFrontendBridge::LocationDeinitialized(void)
+  case RETRO_VFS_FILE_ACCESS_READ_WRITE:
+    return nullptr; // TODO
+  case RETRO_VFS_FILE_ACCESS_UPDATE_EXISTING:
+    return nullptr; // TODO
```
Looks good, only with the two I'm not sure myself whether this is already possible or requires an API update. Think update would make sense to make its use easier to understand.

garbear replied:
Was this functionality added in your VFS work?

AlwinEsch replied:
Maybe it works already, looked at https://github.com/xbmc/xbmc/blob/master/xbmc/platform/posix/filesystem/PosixFile.cpp#L74 and seems with `O_RDWR` for `RETRO_VFS_FILE_ACCESS_READ_WRITE` already supported, also in Windows seems it set.

About `RETRO_VFS_FILE_ACCESS_UPDATE_EXISTING` maybe also works if on second function value the `overwrite` is set to true, only maybe good to compare there between Kodi and Original about inited file positions.
The https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-dev-kit/include/kodi/Filesystem.h#L1877.

garbear replied:
We can open for read and open for write, but can we open for read-write?

AlwinEsch replied:
Yes, the `OpenFileForWrite` should then also have always read access too.

garbear replied:
OK, I've updated the logic for the appropriate flags

AlwinEsch commented on src/libretro/LibretroEnvironment.cpp:
```diff
@@ -540,8 +540,25 @@ bool CLibretroEnvironment::EnvironmentCallback(unsigned int cmd, void *data)
+          CFrontendBridge::WriteFile,
+          CFrontendBridge::FlushFile,
+          CFrontendBridge::RemoveFile,
+          CFrontendBridge::RenameFile,
```
Stay the retro API always in this flow, I think yes only want to confirm?

Seen also that it allow some other functions: https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L1423-L1447
Makes it sense to include them too?

garbear replied:
Yes, this PR only implements version 1. Version 2 & 3 support is planned.

garbear replied:
Version 2 and 3 support has been included in this PR

AlwinEsch reviewed (commented):
Looks good for me.

Hedda commented:
Do you think that this will make it into Kodi's "Matrix" release?

garbear commented:
The clock is ticking down, but we still have a couple months before Matrix's feature freeze. It just depends on time, and this PR is a rather low priority. I'm happy to test and merge if someone finishes the work though.

garbear committed:
Update libretro.h to 2020 version

garbear committed:
Add libretro VFS support (v2 and v3)

garbear commented:
I added support for libretro v2 and v3 VFS API.

The code compiles, but I don't have time to test currently.

@Hedda can you help generate test procedures, like what games and cores should we test?

Hedda commented:
> what games and cores should we test?

@garbear  Do you know where updated info is upstream about which game cores support libretro VFS and version of its VFS?

https://kodi.wiki/view/Game_add-ons#Libretro_cores

Does not look like libretro VFS support and version information is available in all cores today?

https://github.com/libretro/gambatte-libretro/search?q=vfs

Maybe metadata information about VFS and VFS version should be addon.xml for each Kodi Game Addon?

https://github.com/kodi-game/game.libretro.gambatte/tree/master/game.libretro.gambatte

That way the information would be available in downstream and easily accessible to Kodi RetroPlayer end-users for testing?

garbear commented:
> > what games and cores should we test?
> 
> @garbear Do you know where updated info is upstream about which game cores support libretro VFS and version of its VFS?
> 
> https://kodi.wiki/view/Game_add-ons#Libretro_cores

The "VFS support" in the article is something different than this PR - it specifies whether a core can load from memory. As you can see, clearly games that accept entire disc images can't support loading from memory.

The support in this PR is about passing a path, even if "VFS support" is no, to the core, and the core uses Kodi's VFS API.

We can only detect if the libretro core supports loading from memory (VFS support in the core), but we can't detect if the core uses Kodi's VFS API, except by manual code inspection.

> Does not look like libretro VFS support and version information is available in all cores today?
> 
> https://github.com/libretro/gambatte-libretro/search?q=vfs

834 results? https://github.com/search?q=org%3Alibretro+vfs&type=code

> Maybe metadata information about VFS and VFS version should be addon.xml for each Kodi Game Addon?
> 
> https://github.com/kodi-game/game.libretro.gambatte/tree/master/game.libretro.gambatte
> 
> That way the information would be available in downstream and easily accessible to Kodi RetroPlayer end-users for testing?

Manual code inspection just isn't scalable, and has to be re-performed often as cores are updated to use frontend VFS.

garbear commented:
We might as well merge this to allow for testing. @AlwinEsch should I target Matrix because that's what we're developing on? Do we later forward-port everything to Nexus and leave Matrix behind?

garbear committed:
Update libretro.h to 2021 version

Imports libretro.h at version 514534f99b9647a83ee6801bd51ead834a55dac3
of the RetroArch repo.

AlwinEsch commented:
Would be OK, but think that a Matrix and Nexus should remain largely the same. As long as nothing has changed in the API for Kodi.

garbear commented:
So we merge into Matrix?

AlwinEsch reviewed (approved):
Yes can be come in Matrix :+1:
